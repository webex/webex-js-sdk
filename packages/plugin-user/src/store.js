/**!
 *
 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
 */

import {patterns} from '@ciscospark/common';

const usersByEmail = new WeakMap();
const usersById = new WeakMap();

export default class UserStore {
  constructor() {
    usersByEmail.set(this, new Map());
    usersById.set(this, new Map());
  }

  add(user) {
    if (!user.id) {
      return Promise.reject(`\`user.id\` is required`);
    }

    if (!user.emailAddress) {
      return Promise.reject(`\`user.emailAddress\` is required`);
    }

    const p1 = this.getById(user.id)
      .then((u) => usersById.get(this).set(user.id, Object.assign({}, u, user)))
      .catch(() => usersById.get(this).set(user.id, Object.assign({}, user)));

    const p2 = this.getByEmail(user.emailAddress)
      .then((u) => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, u, user)))
      .catch(() => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, user)));

    return Promise.all([p1, p2]);
  }

  get(id) {
    if (patterns.uuid.test(id)) {
      return this.getById(id);
    }

    if (patterns.email.test(id)) {
      return this.getByEmail(id);
    }

    return Promise.reject(`\`id\` does not appear to be a valid user identifier`);
  }

  getById(id) {
    const ret = usersById.get(this).get(id);
    if (ret) {
      return Promise.resolve(ret);
    }
    return Promise.reject(new Error(`No user found identified by specified identifier`));
  }

  getByEmail(email) {
    const ret = usersByEmail.get(this).get(email);
    if (ret) {
      return Promise.resolve(ret);
    }
    return Promise.reject(new Error(`No user found identified by specified identifier`));
  }
}
