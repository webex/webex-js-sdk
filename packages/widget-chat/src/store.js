import {applyMiddleware, combineReducers, compose, createStore} from 'redux';

import thunk from 'redux-thunk';
import createLogger from 'redux-logger';
import {user, conversation} from './reducers';
import sparkReducer from './modules/redux-spark/reducers';

const devtools = window.devToolsExtension || (() => (noop) => noop);

const enhancers = [
  applyMiddleware(
    thunk,
    createLogger({
      level: `info`,
      duration: true,
      collapsed: false,
      logErrors: process.env.NODE_ENV === `production`
    })
  ),
  devtools()
];

export default createStore(
  combineReducers({
    user,
    conversation,
    spark: sparkReducer
  }),
  compose(...enhancers)
);
