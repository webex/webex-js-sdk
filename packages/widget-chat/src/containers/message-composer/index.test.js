/* eslint-disable max-nested-callbacks */
import React from 'react';
import renderer from 'react-test-renderer';
import {Map} from 'immutable';
import {Provider} from 'react-redux';

import store from '../../store.js';
import ConnectedMessageComposer, {MessageComposer} from '.';

let component;

describe(`MessageComposer component`, () => {
  beforeEach(() => {
    component = renderer.create(
      <Provider store={store}>
        <ConnectedMessageComposer
          placeholder="Message Placeholder"
          value="This is a message"
        />
      </Provider>
    );
  });

  describe(`snapshot tests`, () => {
    it(`renders properly`, () => {
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it(`handles value change properly`, () => {
      const value = `abc123`;
      const event = {
        target: {
          value
        }
      };
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      // Not need better way to grab child nodes
      const textarea = tree.children[1].children[0];
      textarea.props.onChange(event);
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe(`class method tests`, () => {
    it(`sends message properly`, () => {
      const props = {
        activity: new Map(),
        submitActivity: jest.fn()
      };
      const messageComposer = new MessageComposer(props);
      messageComposer.handleSubmit();
      expect(messageComposer.props.submitActivity).toHaveBeenCalled();
    });

    describe(`enter key processing`, () => {
      let event;
      const props = {
        activity: new Map()
      };
      const messageComposer = new MessageComposer(props);
      beforeEach(() => {
        messageComposer.handleSubmit = jest.fn();
        event = {
          keyCode: 13,
          shiftKey: false,
          altKey: false,
          ctrlKey: false,
          metaKey: false,
          preventDefault: jest.fn()
        };
      });

      it(`submits upon enter key with no modifiers`, () => {
        messageComposer.handleKeyDown(event);
        expect(messageComposer.handleSubmit).toHaveBeenCalled();
      });

      it(`doesn't submit upon enter key with shift modifiers`, () => {
        event.shiftKey = true;
        messageComposer.handleKeyDown(event);
        expect(messageComposer.handleSubmit).not.toHaveBeenCalled();
      });

      it(`doesn't submit upon enter key with alt modifiers`, () => {
        event.altKey = true;
        messageComposer.handleKeyDown(event);
        expect(messageComposer.handleSubmit).not.toHaveBeenCalled();
      });

      it(`doesn't submit upon enter key with ctrl modifiers`, () => {
        event.ctrlKey = true;
        messageComposer.handleKeyDown(event);
        expect(messageComposer.handleSubmit).not.toHaveBeenCalled();
      });

      it(`doesn't submit upon enter key with meta modifiers`, () => {
        event.metaKey = true;
        messageComposer.handleKeyDown(event);
        expect(messageComposer.handleSubmit).not.toHaveBeenCalled();
      });
    });
  });
});
