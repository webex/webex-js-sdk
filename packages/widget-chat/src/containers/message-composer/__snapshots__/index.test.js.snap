exports[`MessageComposer component snapshot tests handles value change properly 1`] = `
<div
  className="message-composer">
  <textarea
    className="textarea"
    onChange={[Function bound handleChange]}
    onKeyDown={[Function bound handleKeyDown]}
    onSubmit={[Function handleSubmit]}
    placeholder="Message Placeholder"
    value={undefined} />
</div>
`;

exports[`MessageComposer component snapshot tests handles value change properly 2`] = `
<div
  className="message-composer">
  <textarea
    className="textarea"
    onChange={[Function bound handleChange]}
    onKeyDown={[Function bound handleKeyDown]}
    onSubmit={[Function handleSubmit]}
    placeholder="Message Placeholder"
    value="abc123" />
</div>
`;

exports[`MessageComposer component snapshot tests renders properly 1`] = `
<div
  className="message-composer">
  <textarea
    className="textarea"
    onChange={[Function bound handleChange]}
    onKeyDown={[Function bound handleKeyDown]}
    onSubmit={[Function handleSubmit]}
    placeholder="Message Placeholder"
    value={undefined} />
</div>
`;
